File: scope.go
package groundcover

import (
	"context"
	"sync"
)

// Scope holds request-level data merged into every event captured with the
// owning context. It sits between global defaults and per-call options in the
// merge precedence.
//
// A Scope is mutable and safe for concurrent use. Middleware installs one fresh,
// isolated Scope per request (see WithIsolatedScope); handlers then mutate that
// same Scope through SetUser / WithScope, and the captured event observes those
// changes without the handler having to thread a new context back.
type Scope struct {
	mu          sync.Mutex
	user        User
	attributes  Attributes
	level       Level
	fingerprint string
	sessionID   string
	anonymousID string
}

// SetUser sets the identity on the scope.
func (s *Scope) SetUser(u User) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.user = u
}

// SetAttributes merges attributes into the scope.
func (s *Scope) SetAttributes(a Attributes) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.attributes == nil {
		s.attributes = make(Attributes, len(a))
	}
	s.attributes.merge(a)
}

// SetAttribute sets a single attribute on the scope.
func (s *Scope) SetAttribute(key string, value any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.attributes == nil {
		s.attributes = make(Attributes, 1)
	}
	s.attributes[key] = value
}

// SetLevel sets the default severity for events in this scope. Note that the
// scope level never downgrades an intrinsically-fatal event (a recovered panic).
func (s *Scope) SetLevel(l Level) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.level = l
}

// SetFingerprint overrides the grouping fingerprint for events in this scope.
func (s *Scope) SetFingerprint(fp string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.fingerprint = fp
}

// SetSessionID sets the session identifier for events in this scope.
func (s *Scope) SetSessionID(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.sessionID = id
}

// SetAnonymousID sets the pre-auth anonymous identifier for events in this scope.
func (s *Scope) SetAnonymousID(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.anonymousID = id
}

// clone returns a deep copy of the scope (without the source's lock state).
func (s *Scope) clone() *Scope {
	if s == nil {
		return &Scope{}
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	return &Scope{
		user:        s.user,
		attributes:  s.attributes.clone(),
		level:       s.level,
		fingerprint: s.fingerprint,
		sessionID:   s.sessionID,
		anonymousID: s.anonymousID,
	}
}

// applyTo merges the scope into an event. It runs after global defaults and
// before per-call options.
func (s *Scope) applyTo(e *Event) {
	if s == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.user.isZero() {
		e.User = s.user
	}
	if len(s.attributes) > 0 {
		if e.Attributes == nil {
			e.Attributes = make(Attributes, len(s.attributes))
		}
		e.Attributes.merge(s.attributes)
	}
	// A scope level fills in / overrides the default, but must never downgrade
	// an intrinsically-fatal event (a recovered panic).
	if s.level.valid() && !e.levelLocked {
		e.Level = s.level
	}
	if s.fingerprint != "" {
		e.Fingerprint = s.fingerprint
	}
	if s.sessionID != "" {
		e.SessionID = s.sessionID
	}
	if s.anonymousID != "" {
		e.AnonymousID = s.anonymousID
	}
}

// scopeKey is the context key type for the request scope.
type scopeKey struct{}

// scopeFromContext returns the scope stored in ctx, or nil if none.
func scopeFromContext(ctx context.Context) *Scope {
	if ctx == nil {
		return nil
	}
	sc, _ := ctx.Value(scopeKey{}).(*Scope)
	return sc
}

// contextWithScope returns a derived context carrying sc.
func contextWithScope(ctx context.Context, sc *Scope) context.Context {
	if ctx == nil {
		ctx = context.Background()
	}
	return context.WithValue(ctx, scopeKey{}, sc)
}

// ensureScope returns the scope already attached to ctx (so mutations are
// visible to whoever else holds the context), creating and attaching a fresh one
// if there is none.
func ensureScope(ctx context.Context) (context.Context, *Scope) {
	if sc := scopeFromContext(ctx); sc != nil {
		return ctx, sc
	}
	sc := &Scope{}
	return contextWithScope(ctx, sc), sc
}

// isolatedScopeContext returns a context carrying a fresh, isolated copy of the
// current scope. Used at request boundaries so per-request mutations never leak
// across requests.
func isolatedScopeContext(ctx context.Context) context.Context {
	return contextWithScope(ctx, scopeFromContext(ctx).clone())
}

File: groundcover-python/src/groundcover/_scope.py
"""Request-level scope, carried on contextvars.

The Go SDK threads scope through ``context.Context``; the Python translation
stores it in a ``contextvars.ContextVar``, which propagates naturally through
async tasks and can be copied into threads with ``contextvars.copy_context``.
"""

from __future__ import annotations

import contextlib
import contextvars
import threading
from typing import Any, Iterator, Optional, Union

from ._event import Attributes, Event
from ._level import Level, coerce_level
from ._user import User


class Scope:
    """Holds request-level data merged into every event captured while the
    scope is current. It sits between global defaults and per-call options in
    the merge precedence.

    A Scope is mutable and safe for concurrent use. Middleware installs one
    fresh, isolated Scope per request (see isolated_scope()); handlers then
    mutate that same Scope through set_user / with_scope, and the captured
    event observes those changes.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._user = User()
        self._attributes: Attributes = {}
        self._level: Optional[Level] = None
        self._fingerprint = ""
        self._session_id = ""
        self._anonymous_id = ""

    def set_user(self, user: User) -> None:
        """Set the identity on the scope."""
        with self._lock:
            self._user = user.copy()

    def set_attributes(self, attributes: Attributes) -> None:
        """Merge attributes into the scope."""
        with self._lock:
            self._attributes.update(attributes)

    def set_attribute(self, key: str, value: Any) -> None:
        """Set a single attribute on the scope."""
        with self._lock:
            self._attributes[key] = value

    def set_level(self, level: Union[Level, str]) -> None:
        """Set the default severity for events in this scope. Note that the
        scope level never downgrades an intrinsically-fatal event (an uncaught
        exception)."""
        with self._lock:
            self._level = coerce_level(level)

    def set_fingerprint(self, fingerprint: str) -> None:
        """Override the grouping fingerprint for events in this scope."""
        with self._lock:
            self._fingerprint = fingerprint

    def set_session_id(self, session_id: str) -> None:
        """Set the session identifier for events in this scope."""
        with self._lock:
            self._session_id = session_id

    def set_anonymous_id(self, anonymous_id: str) -> None:
        """Set the pre-auth anonymous identifier for events in this scope."""
        with self._lock:
            self._anonymous_id = anonymous_id

    def clone(self) -> Scope:
        """Return a deep copy of the scope."""
        out = Scope()
        with self._lock:
            out._user = self._user.copy()
            out._attributes = dict(self._attributes)
            out._level = self._level
            out._fingerprint = self._fingerprint
            out._session_id = self._session_id
            out._anonymous_id = self._anonymous_id
        return out

    def apply_to(self, e: Event) -> None:
        """Merge the scope into an event. It runs after global defaults and
        before per-call options."""
        with self._lock:
            if not self._user.is_zero():
                e.user = self._user.copy()
            if self._attributes:
                e.attributes.update(self._attributes)
            # A scope level fills in / overrides the default, but must never
            # downgrade an intrinsically-fatal event (an uncaught exception).
            if self._level is not None and not e._level_locked:
                e.level = self._level
            if self._fingerprint:
                e.fingerprint = self._fingerprint
            if self._session_id:
                e.session_id = self._session_id
            if self._anonymous_id:
                e.anonymous_id = self._anonymous_id


_scope_var: contextvars.ContextVar[Optional[Scope]] = contextvars.ContextVar(
    "groundcover_scope", default=None
)


def current_scope() -> Optional[Scope]:
    """Return the scope attached to the current context, or None."""
    return _scope_var.get()


def ensure_scope() -> Scope:
    """Return the scope already attached to the current context (so mutations
    are visible to everything sharing that context), creating and attaching a
    fresh one if there is none."""
    sc = _scope_var.get()
    if sc is not None:
        return sc
    sc = Scope()
    _scope_var.set(sc)
    return sc


@contextlib.contextmanager
def isolated_scope() -> Iterator[Scope]:
    """Context manager installing a fresh, isolated copy of the current scope.
    Middleware uses it at request boundaries so per-request identity and
    attributes never leak across requests."""
    current = _scope_var.get()
    sc = current.clone() if current is not None else Scope()
    token = _scope_var.set(sc)
    try:
        yield sc
    finally:
        _scope_var.reset(token)
